File: day2/solution2.2.ts
import * as fs from "fs";

function isSafeReport(report: number[]): boolean {
    // Calculate the differences between adjacent levels
    const differences = report.slice(1).map((level, i) => level - report[i]);

    // Check if all differences are valid (between 1 and 3 or -1 and -3)
    if (!differences.every(diff => Math.abs(diff) >= 1 && Math.abs(diff) <= 3)) {
        return false;
    }

    // Check if the report is strictly increasing or strictly decreasing
    const allIncreasing = differences.every(diff => diff > 0);
    const allDecreasing = differences.every(diff => diff < 0);

    return allIncreasing || allDecreasing;
}

function isSafeWithDampener(report: number[]): boolean {
    if (isSafeReport(report)) {
        return true;
    }

    // Try removing each level and check if the modified report is safe
    for (let i = 0; i < report.length; i++) {
        const modifiedReport = report.slice(0, i).concat(report.slice(i + 1));
        if (isSafeReport(modifiedReport)) {
            return true;
        }
    }

    return false;
}

function countSafeReportsWithDampener(filePath: string): number {
    // Read and parse the input file
    const data = fs.readFileSync(filePath, "utf-8");
    const lines = data.trim().split("\n");

    let safeCount = 0;
    for (const line of lines) {
        const report = line.trim().split(" ").map(Number);
        if (isSafeWithDampener(report)) {
            safeCount++;
        }
    }

    return safeCount;
}

// Main Execution
const inputFile = "./input.txt";
const safeReports = countSafeReportsWithDampener(inputFile);
console.log(`Number of safe reports: ${safeReports}`);

File: day2/solution2.2.py
def is_safe_report(report):
    """
    Determine if a report is safe without modifications.
    """
    differences = [report[i + 1] - report[i] for i in range(len(report) - 1)]

    # Check if all differences are valid (between 1 and 3 or -1 and -3)
    if not all(1 <= abs(diff) <= 3 for diff in differences):
        return False

    # Check if the report is strictly increasing or strictly decreasing
    return all(diff > 0 for diff in differences) or all(diff < 0 for diff in differences)


def is_safe_with_dampener(report):
    """
    Determine if a report can be made safe by removing one level.
    """
    if is_safe_report(report):
        return True

    # Try removing each level (except the first and last)
    for i in range(len(report)):
        modified_report = report[:i] + report[i + 1:]
        if is_safe_report(modified_report):
            return True

    return False


def count_safe_reports_with_dampener(file_path):
    """
    Count the number of safe reports considering the Problem Dampener.
    """
    with open(file_path, 'r') as file:
        lines = file.readlines()

    safe_count = 0
    for line in lines:
        report = list(map(int, line.strip().split()))
        if is_safe_with_dampener(report):
            safe_count += 1

    return safe_count


if __name__ == "__main__":
    input_file = "./input.txt"
    safe_reports = count_safe_reports_with_dampener(input_file)
    print(f"Number of safe reports: {safe_reports}")